eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentRequest.scala

```diff
@@ -487,6 +487,47 @@ object PaymentRequest {
     )
   }
 
+  private def readBoltData(input: String): Bolt11Data = {
+    val lowercaseInput = input.toLowerCase
+    val separatorIndex = lowercaseInput.lastIndexOf('1')
+    val hrp = lowercaseInput.take(separatorIndex)
+    val prefix: String = prefixes.values.find(prefix => hrp.startsWith(prefix)).getOrElse(throw new RuntimeException("unknown prefix"))
+    val data = string2Bits(lowercaseInput.slice(separatorIndex + 1, lowercaseInput.length - 6)) // 6 == checksum size
+    Codecs.bolt11DataCodec.decode(data).require.value
+  }
+
+  /**
+   * Extracts the description from a serialized payment request that is **expected to be valid**.
+   * Throws an error if the payment request is not valid.
+   *
+   * @param input valid serialized payment request
+   * @return description as a String. If the description is a hash, returns the hash value as a String.
+   */
+  def fastReadDescription(input: String): String = {
```

Comment:
If you're calling both `fastReadDescription` and `fastHasExpired` for the same invoice, you'll be decoding the data twice...if that's the case these methods should take the decoded data as input instead of the string?

Reply:
These two methods are indeed called twice. 

However `fastHasExpired` is called very early, in `listPaymentsOverview` by `buildIncomingPaymentStatus` to get the status of an incoming payment. Note that it is also used by the regular `xxxIncomingPayment` methods so changing this method's signature would have some side-effects.

`fastReadDescription` is called at the last moment when actually displaying the description from the payment request (so it's only called for the currently visible payments, and not called for payments that are in the list but not displayed).

Also, AFAICT on Android the slow part in the `PaymentRequest.read` is not reading the bolt11 data but the recovering the public key.

Reply:
Ok makes sense, we can re-visit that later if we want to improve the perf even more :)